.github/workflows/test.yml

```diff
@@ -0,0 +1,321 @@
+            sudo apt-get update
+            sudo apt-get install --yes \
+              build-essential \
+              curl \
+              libcurl4-openssl-dev \
+              libssl-dev
+      - uses: actions/setup-node@v2-beta
+        with:
+          node-version: '14'
+      - name: Run webpack to build static assets
+        shell: bash
+        run: |
+            npm install
+            npm run webpack
+      - name: Setup Python package dependencies
+        shell: bash
+        run: |
+            # Install development requirements, package requirements, and frozen image
+            # requirements, with an increasing priority.
+            pip install -r dev-requirements.txt
+            pip install .
+            pip install -r helm-chart/images/binderhub/requirements.txt
+      - name: Use chartpress to create the helm chart
+        shell: bash
+        run: |
+            # Use chartpress to create the helm chart and build its images
+            helm dependency update ./helm-chart/binderhub
+            (cd helm-chart && chartpress)
+            git --no-pager diff
+      - name: Validate the chart against the k8s API
+        shell: bash
+        run: |
+            helm template --validate binderhub-test helm-chart/binderhub \
+                --values testing/k8s-binder-k8s-hub/binderhub-chart-config.yaml \
+                --set config.BinderHub.hub_url=http://localhost:30902 \
+                --set config.BinderHub.access_token=$GITHUB_ACCESS_TOKEN
+      - name: Install the chart
+        shell: bash
+        run: |
+          helm upgrade --install binderhub-test helm-chart/binderhub \
+              --values testing/k8s-binder-k8s-hub/binderhub-chart-config.yaml \
+              --set config.BinderHub.hub_url=http://loaclhost:30902 \
+              --set config.BinderHub.access_token=$GITHUB_ACCESS_TOKEN
+      - name: Wait for BinderHub to be ready
+        shell: bash
+        run: |
+            # Wait for JupyterHub and BinderHub to be ready
+            set -e
+            . ci/common
+            await_jupyterhub
+            await_binderhub
+            # Next few commands are here for sanity checking/debugging of CI
+            # related issues
+            echo "Proxy service details"
+            kubectl get svc proxy-public -o json
+            echo
+            echo "kubectl describe svc proxy-public"
+            kubectl describe svc proxy-public
+            echo
+            echo "kubectl get svc"
+            kubectl get svc
+            echo
+            echo "curl http://localhost:30902/hub/api/" should print the JupyterHub version
+            curl http://localhost:30902/hub/api/
+            echo
+            echo "curl http://localhost:30901/health" to check BinderHub\'s health
+            curl http://localhost:30901/health
+      - name: Run tests
+        shell: bash
+        run: |
+            export BINDER_URL=http://localhost:30901
+            pytest -m "remote" -vx --cov binderhub
```

Comment:
I'd love to see these three jobs (main, auth, helm) be combined into one job template that is parameterized into three different jobs.

I learned some tricks from @manics such as using `${{ matrix.my_setting }}` to conditionally do things in steps or conditionally run entire steps. This PR contains some examples of this: https://github.com/jupyterhub/jupyterhub/pull/3246.

Reply:
Yes it would be nice to have one job and use a matrix but I am not sure that having lots of `if` statements to conditionally enable/disable steps based on the value of the matrix flags is worth it. Having three separate jobs makes it very explicit what steps should be executed when. So for now I'd like to keep it like this instead of adding ga layer of abstraction

Reply:
> Yes it would be nice to have one job and use a matrix but I am not sure that having lots of if statements to conditionally enable/disable steps based on the value of the matrix flags is worth it. Having three separate jobs makes it very explicit what steps should be executed when. So for now I'd like to keep it like this instead of adding ga layer of abstraction

I agree things become more explicit, but like assembler which is also more explicit, it is also harder to overview and maintain. It is a quite extreme comparison and this case isn't at all that extreme, but I feel quite strongly about it, it relates to my time investment in #ci I would like to see them combined again, I put in a lot of thorough work in https://github.com/jupyterhub/binderhub/pull/1188 which included gathering all the differences into a few steps which were separated from the common.

Here is the current `.travis.yml` file with its different steps for the three test scenarios, the common scenarios are defined in `before_install` and `install`. I suggest that we do the same thing here, we have a set of common steps and then we have three test scenarios in three steps that we activate conditionally.


```yaml
    - stage: test
      name: main
      script:
        - ./testing/local-binder-k8s-hub/install-jupyterhub-chart
        - await_jupyterhub
        - pytest -m "not auth" -vx --cov binderhub

    - stage: test
      name: auth
      script:
        - ./testing/local-binder-k8s-hub/install-jupyterhub-chart --auth
        - await_jupyterhub
        - pytest -m "auth" -vx --cov binderhub

    - stage: test
      name: helm
      script:
        # Prepare:  builds images and updates default values.yaml to reference
        #           the images
        - echo "Preparing local BinderHub Helm chart"
        - helm dependency update ./helm-chart/binderhub
        - (cd helm-chart && chartpress)
        - git --no-pager diff

        # Validate: renders Helm chart templates and validates the rendered k8s
        #           manifests against the k8s api-server.
        - |
          helm template --validate binderhub-test helm-chart/binderhub \
              --values testing/k8s-binder-k8s-hub/binderhub-chart-config.yaml \
              --set config.BinderHub.hub_url=http://$(minikube ip):30902 \
              --set config.BinderHub.access_token=$GITHUB_ACCESS_TOKEN

        # Install:  installs the local BinderHub Helm chart and await it to
        #           become ready.
        - echo "Installing local BinderHub Helm chart with JupyterHub as a dependency into k8s cluster"
        - |
          helm upgrade --install binderhub-test helm-chart/binderhub \
              --values testing/k8s-binder-k8s-hub/binderhub-chart-config.yaml \
              --set config.BinderHub.hub_url=http://$(minikube ip):30902 \
              --set config.BinderHub.access_token=$GITHUB_ACCESS_TOKEN
        - await_binderhub

        # Test
        - export BINDER_URL=http://$(minikube ip):30901
        - echo "Running tests against $BINDER_URL"
        - pytest -m "remote" -vx --cov binderhub
```

Reply:
In travis it was nice because we could share all the setup steps and then have steps that were different. The only way I know to share steps between jobs in GH Actions is to make steps conditional.

This is now implemented but I don't think it is easy to read because you have to check the `if:` for every step. However I don't have a strong enough opinion on this so will just do what ever others want.

Reply:
A compromise could be to move the common functionality into new bash functions in `ci/common`, and keep the three separate jobs?

Reply:
Thank you @betatim :heart:

@manics I think we can extract some bash functions, but I think we should let the github workflow be easy to parse in form of "await_jupyterhub" or similar, rather than "do_everything_script". I believe we already have extracted most relevant parts though.